### ISSUE TITLE
Straighten UVs

### ISSUE THREAD

**Issue by Roland09** [nutti/Magic-UV#56]
It would be great if Magic-UV could do something similar what UV-Squares does, but in a more improving way: Select uv vertices and try to align and flatten them along the x and y coordinates.

Example:

* add a cylinder without top and bottom face
* create 2 loop cuts
* set selection to face select
* select the middle ring and extrude it along x and y

Now when you uv-unwrap it, you don't get a flat uv grid anymore as you would have with a simple cylinder. It's distorted, even though it still is a grid. It would be great to align the uv vertices automatically along the x and y axis, similar to what UV-Squares does.

The improvement would be that e. g. the alignment is only performed on the selected vertices.

Example screenshot: top is what you get after unwrapping, bottom is how it should be after you straighten the uvs

![straighten](https://user-images.githubusercontent.com/10963432/27760859-ce6e1c68-5e51-11e7-8d3a-eeb98af93634.png)

This example is more complex and hardly to be automated. However, this is what I mean by only straighten the selected uvs. Example object:

* add a cylinder without top and bottom face
* create 2 loop cuts
* set selection to face select
* select the middle ring, checker deselect all and extrude indivisually along x and y

Unwrapping gives you this:

![straighten 02](https://user-images.githubusercontent.com/10963432/27760903-af834b74-5e52-11e7-8148-2e3e51014c77.png)

It would at least be helpful to select the top and bottom rows individually and align them.

**Comment by nutti** [comment]
thanks for your request, and sorry for my late response.

does this request mean improvement of UV-squares?

**Comment by WeCode4Fun** [comment]
No, I thought of it as a feature request for Magic-UV. Especially now that
Magic-UV is part of Blender. UV-squares does a good initial job, but it
lacks special features like e. g. straighten only the selected vertices. I
think Magic-UV is already awesome and this could be a great and useful
enhancement. Also, UV-squares doesn't seem to be developed anymore. If this
is not what Magic-UV is intended for, please ignore my request.

**Comment by nutti** [comment]
@WeCode4Fun 

yes, i think your request makes Magic UV more powerful.

i would like to check this feature's requirement.

in the last case you mentioned on the first posts in this issue, it seems that top and bottom Coordinate V is inclined.
So, do you want align them?

**Comment by WeCode4Fun** [comment]
Yes, they should be aligned. However I rather think a simple approach could work for this as well;

* align the 1st line horizontally
* iterate through every vertex of the other rows from top to bottom and align them depending on the reference vertex of the first row.

But only selected vertices. This is what e. g. UV Squares can't do. It can't align only selected vertices. It can't align the grid in the 2nd image of the 1st post at all. I don't think there could be a good automatism for the case I posted anyway (I don't know), but it would at least be helpful to have some kind of grid on which you can manually align the rest of the vertices which can't be aligned via automatism.

Basically think of it to work like a curtain. You hang it at the top row and let gravity do its work. In the end what you get would be a grid.

Currently all the other in-built uv tools seem to allow only alignment of 1 single selected row. And that's cumbersome.

**Comment by schroef** [comment]
@WeCode4Fun 
Thats exactly how UV_tool works. Its a very usefull addon. Also has some other options for smoothing and relaxing UV sets.

https://blenderartists.org/forum/showthread.php?294904-AddOn-UV_Tool

**Comment by WeCode4Fun** [comment]
@schroef I know that one. It's buggy, not intuitive, can't deal with selected uv vertices only, requires python 3.6 on windows whereas blender comes with 3.5. And most of all, if one wants to fix things themselves, the code is in chinese, even the variables.

But as I said, if this is not what Magic UV is designed for, please ignore my request and close this.

**Comment by schroef** [comment]
I cant remember how, but works fine for me on OSX. I think i upgraded python but 2.78 in the lib still has 3.5. I dont have issues. Sometimes and it spits out these Chinese errors indeed. Perhaps a google translate could put it in English, the coder isnt very good at English i believe.

Not sure what you mean by it tries to fix things, it works just fine for straightening meshes. There hardly plugins what this one can do with straightening uv maps. Not that i know of, there is Quad Unwrap or so. But most of the time rips meshes up.

Indeed there is a issue that it loops hrough all vertices in the mesh. I think that could be fixed, but i cant look in that code.

**Comment by WeCode4Fun** [comment]
I meant if you try to fix the code yourself. Here's an example code line:

f偏移比例=f此点uv实减应长十/Lf每段UV实长十Now[i序应];#比例 决定uv应偏移量 #一定在原来方向上偏移

**Comment by nutti** [comment]
@WeCode4Fun 

hi, is it ok to close this issue?
i thought that your requested feature is not included in UV_Tool.

Magic UV have included other old UV add-on, such as World Scale UV, Move UV from 3D View, etc...
so, there is no design about Magic UV :)

if you want, i'll try to check his script (UV_Tool) and includes it. but it will take long time because i'm busy now and his script is difficult to read :(

and..., it seems it is difficult to include his script, because his add-on has no license (no GPL, and no MIT).
if you want to include his script to Magic UV, i must agree with him.

anyway, thanks for your request!!

**Comment by WeCode4Fun** [comment]
@nutti Well, no it's not. But from the discussion it appeared to me that Magic UV isn't the right place for this request. I reopen it and let you guys decide about the closing.

Of course it would be super awesome to have this feature integrated. Also because Magic UV is part of Blender now and UV_Tool would be a very valuable addtion. Same is valid for UV Squares by the way. Less, code, very readable. I had no idea that combining other repositories would be an option.

I'm not sure though if integrating the chinese code of UV_Tool is easy. Also, there need to be some fixes. eg the properties of UV_Tool are in the panel of the 3D View instead of the Image view in Blender. Hence requires other fixes as well. It sounds like a lot of work.

**Comment by schroef** [comment]
Thats weird, for me uv_tool is in the image editor.
<img width="468" alt="screen shot 2017-07-11 at 4 49 10 pm" src="https://user-images.githubusercontent.com/6923008/28090044-ffa23d50-6658-11e7-90d3-06e13d17a716.png">

**Comment by WeCode4Fun** [comment]
@schroef The properties of the tool, not the tool itself. Do you have 3d view also open? I have a split view: left is uv/image editor, right is 3d view. When you hit the "Align" button you get the LMR option in the 3d view:

![lmr](https://user-images.githubusercontent.com/10963432/28102747-9bcbe118-66d1-11e7-9b7f-f190a7d981a9.png)

**Comment by schroef** [comment]
Aha sorry i see yeah, ive so many panels its always squashed to the bottom. Never noticed it. Tried checking in the code, but the part it spits out is also in Chinese. Its quite weird thought cause most of the code in that section belong to align button, which is set to the IMAGE_EDITOR.

**Comment by nutti** [comment]
@Roland09 

Hi,

it's been while I posted here.
sorry for my tardiness for late response.

I'm now talking to him who is developer of UV_Tools, and he considers positively about including his script to Magic UV.
I will continue talking him about future works if his add-on will be merged.

https://blenderartists.org/forum/showthread.php?294904-AddOn-UV_Tool&p=3255986&viewfull=1#post3255986

**Comment by Roland09** [comment]
That's awesome. Thank you very much for your efforts!!! :-)

**Comment by nutti** [comment]
imdjs is rewriting his scripts to English now.
so please wait until he finished his work.

And it will become big change about Magic UV.
So i will release Magic UV v4.5 at first, then i will merge UV_Tools and may change current UI.

**Comment by Roland09** [comment]
Great! imo it will become a huge benefit for Blender users :-)

**Comment by nutti** [comment]
I started English translation in forked repository.

https://github.com/nutti/UV_Tool

please see above links for the merging process.

**Comment by nutti** [comment]
@Roland09 

Hi

I have already cleaned up 80% of UV_Tools and merged it into Magic UV.
Can you try merged version?

<img width="220" alt="merge_uv_tools" src="https://user-images.githubusercontent.com/1463952/33523245-16b0439e-d845-11e7-88f6-4334415b3624.png">


you can download trial version from below link.

https://github.com/nutti/Magic-UV/archive/issue_56.zip

**Comment by Roland09** [comment]
Hi,

thank you!!! I tried, but I get the following error when I select e. g.
multiple uv vertices along a line and hit "straighten":

Error: "Isolated UVs are found  (Expected 2 but 22").

Any idea what it means? I attached the .blend file I was using.

Thank you very much for your efforts!

Cheers
Roland

**Comment by nutti** [comment]
@Roland09 

Sorry, i misunderstood original feature.
So i will rewrite these scripts again.

please wait...

**Comment by nutti** [comment]
@Roland09 

Sorry for my late work.
I finished to reimplement all features in UV_Tool.

they are already included in develop branch.

https://github.com/nutti/Magic-UV/archive/develop.zip

can you try it?
if there is no problem, I will release new version of Magic UV.

**Comment by Roland09** [comment]
Thank you for your efforts! It appears there is an endless loop, Blender
freezes. Reproduction:

* open the attached test.blend file, all vertices are selected in UV editor
* select the Magic UV tab and hit Smooth

**Comment by nutti** [comment]
@Roland09 

Hi, thanks for your testing.
Where is your attached file.
I can not find any file in GitHub and E-mail.

**Comment by Roland09** [comment]
@nutti I added it to this post apparently it doesn't get sent when one replies via mail

It's just a cylinder with a few loop cuts and a seam along the vertical axis.

[test.zip](https://github.com/nutti/Magic-UV/files/1572846/test.zip)

**Comment by nutti** [comment]
@Roland09 

Thanks.
I tried your procedure.

I think you mistake about the usage.
You must select only one edge as the picture showed in below.

![github_20171220](https://user-images.githubusercontent.com/1463952/34207664-c355adc8-e5ce-11e7-8f25-100f39a72c11.png)

**Comment by Roland09** [comment]
Yes, but Blender shouldn't freeze under any condition, otherwise all your
work is gone :-)

**Comment by nutti** [comment]
@Roland09 

yes, I think too.
but I can not reproduce your error.
how can I reproduce?

**Comment by nutti** [comment]
@Roland09 

Sorry, I can reproduce your error.
And I fixed it already.

please check it.

https://github.com/nutti/Magic-UV/archive/develop.zip

**Comment by Roland09** [comment]
Awesome, thanks a lot! It's great :-)

A question: Having the properties in the Tool Shelf is a Blender feature, isn't it? Because if you have only the UV Editing view opened, you won't see the properties unless you open the 3D View as well. Example (red bordered ara):

![properties](https://user-images.githubusercontent.com/10963432/34373919-6acadd94-eade-11e7-8f52-a73bcbeda008.png)

**Comment by schroef** [comment]
If its a function for in the 3dview it makes sense, if its a UV editor function than thats a mistake :)

**Comment by nutti** [comment]
@Roland09 @schroef 

Yes, I think it is not good UI.
So I changed the UI dramatically in next version.

Here, this is a trial version.

<img width="1245" alt="ba_20171230" src="https://user-images.githubusercontent.com/1463952/34450448-23e00ede-ed4c-11e7-923a-0ee937410237.png">

https://github.com/nutti/Magic-UV/archive/develop.zip

please check it.

**Comment by Roland09** [comment]
The properties are still visible in the Tool Shelf in addition, but that's okay imo. The difference I notice though:

* in the Tool Shelf you can change the properties after you pressed e. g. "Straighten" and they are still applied
* in "Align UV" at UV/Image Editor the properties are applied only once, modification of the properties has no effect unless you press "Striaghten" again

**Comment by nutti** [comment]
@Roland09 

> The properties are still visible in the Tool Shelf in addition, but that's okay imo.

I can hide this option in View3D.
But I deliberately remained this options because you can retry the operation with different options.

> in the Tool Shelf you can change the properties after you pressed e. g. "Straighten" and they are still applied
> in "Align UV" at UV/Image Editor the properties are applied only once, modification of the properties has no effect unless you press "Striaghten" again

Yes, this is an expected behavior.
Original UV_Tool can only do latter case.
As I mentioned before, I deliberately remain this options for retrying the operation with different options.
So, Magic UV can do former case too.

**Comment by nutti** [comment]
I think this issue is already cleared.
If you have any other issues about this feature, please make new issue or reopen this issue.

Thanks for your cooperation.
